---
title: "R Notebook"
output: html_notebook
---

Install Giotto, GiottoData and the Giotto environmment
```{r Load required libraries}
if(!"Giotto" %in% installed.packages()) {
  devtools::install_github("drieslab/Giotto@suite", force = TRUE)
}
library("Giotto")

if(!"GiottoData" %in% installed.packages()) {
  remotes::install_github("drieslab/GiottoData")
}
library(GiottoData)

genv_exists <- checkGiottoEnvironment()
if(!genv_exists){
  # The following command need only be run once to install the Giotto environment.
  installGiottoEnvironment(force_environment = TRUE)
  
# library("tidyverse")
}
```

## Create the Giotto object

```{r Set up work space}
results_folder = "Giotto_Results"

instrs = createGiottoInstructions(save_dir = results_folder,
                                  save_plot = TRUE,
                                  show_plot = TRUE)
```

```{r create Giotto object}
## provide path to visium folder
data_path<- "/Users/graham/Documents/Polyomics/Pr2029/run1/D1/outs/"

## directly from visium folder
visium_giotto_object <- createGiottoVisiumObject(visium_dir = data_path,
                                        expr_data = 'raw',
                                        png_name = 'tissue_lowres_image.png',
                                        gene_column_index = 2,
                                        instructions = instrs,
                                        do_manual_adj = TRUE,
                                        xmax_adj = 248,
                                        xmin_adj = 365,
                                        ymax_adj = 460,
                                        ymin_adj = 200)

# Show images associated with Giotto object
showGiottoImageNames(visium_giotto_object)

# Check metadata
pDataDT(visium_giotto_object)

## show plot
slide_image <- spatPlot2D(gobject = visium_giotto_object, 
                          cell_color = 'in_tissue',
                          point_size = 2,
                          cell_color_code = c('0' = 'lightgrey', '1' = 'blue'),
                          show_image = T, 
                          image_name = 'image')

slide_image
```

# Process Giotto Visium object
## Filter
Filters are based on:
**expression_threshold** - set the minimum expression level
**feat_det_in_min_cells** - set the minimum number of cells that expresses a feature/gene
**min_det_feats_per_cell** - set the minimum number of features/genes expressed per cell


```{r cell distribution}
filterDistributions(visium_giotto_object, detection = 'cells')
```

```{r feature distribution}
filterDistributions(visium_giotto_object, detection = 'feats')
```
This is a visualisation showing the number of features lost based on the various filtering parameters
```{r}
filterCombinations(gobject = visium_giotto_object, 
                   expression_thresholds = c(1, 2, 3),
                   feat_det_in_min_cells = c(2, 5, 10), 
                   min_det_feats_per_cell = c(100, 200, 500))
```

Use the plots above to inform the cutoff parameters for filterGiotto.
In this example, an expression value of 1, feat_det_in_min_cells of 10 and min_det_feats_per_cell of 200.
Also, filter out spot not covered by the tissue.
```{r filter}
## subset on spots that were covered by tissue
metadata <- pDataDT(visium_giotto_object)
in_tissue_barcodes <- metadata[in_tissue == 1]$cell_ID
visium_giotto_object <- subsetGiotto(visium_giotto_object, cell_ids = in_tissue_barcodes)

visium_giotto_object <- filterGiotto(gobject = visium_giotto_object,
                             expression_threshold = 1,
                             feat_det_in_min_cells = 10,
                             min_det_feats_per_cell = 200,
                             expression_values = c('raw'),
                             verbose = T)
```

## Normalize the expression matrix
There are three normalisation algorythms:
**Standard** - Normailses the data by library size and a custom scale factor, then log transforms and z-scores the data by genes or cells, which is specified by scale_order. The scale_order options are first_feats or first_cells.
```{r normalise standard}
visium_giotto_object <- normalizeGiotto(gobject = visium_giotto_object, 
                                norm_methods = "standard", 
                                scale_feats = TRUE,
                                scalefactor = 6000, 
                                scale_order = "first_feats",
                                verbose =TRUE)
```

**pearson_resid** - Expected values calculated based on Pearson correlations, z-scores are then calculated based on the observed values and the calculated expected values.
```{r normalise pearson-resid}
visium_giotto_object <- normalizeGiotto(gobject = visium_giotto_object, 
                                norm_methods = "pearson_resid", 
                                scale_feats = TRUE,
                                scalefactor = 6000, 
                                scale_order = "first_feats",
                                verbose =TRUE)
```

**osmFISH** - This normalisation method is for *in situ* RNA data using a method described in Codeluppi *et al.* Feature/gene counts are normalised individually by the total feature/gene count and then multiplied by the total number of features/genes. Then, cells are individually normalised by dividing the normalised feature counts by the total feature counts per cell, then scaled by the total number of cells.
This method is not shown as the data is unlikely to be from *in situ* RNA experiments. However, to use the algorithm can specify "osmFISH" as the argument for norm_methods.
```{r normalise osmFISH}
# visium_giotto_object <- normalizeGiotto(gobject = visium_giotto_object, 
#                                 norm_methods = "osmFISH", 
#                                 scale_feats = TRUE,
#                                 scalefactor = 6000, 
#                                 verbose =TRUE)
```


## Add gene & cell statistics and metadata
The function addStatistics will add the following statistics to cell metadata:
**nr_feats** - Denotes how many features are detected per cell
**perc_feats** - Denotes the percentage of features detected per cell
**total_expr** - Shows the total sum of feature expression per cell

It will also add the following statistics to feature metadata:
**nr_cells** - Denotes how many cells in which the feature is detected
**per_cells** - Denotes the percentage of cells in which the feature is detected
**total_expr** - Shows the total sum of feature expression in all cells
**mean_expr** - Average feature expression in all cells
**mean_expr_det** - Average feature expression in cells with detectable levels of the feature

```{r Add statistics}
visium_giotto_object <- addStatistics(gobject = visium_giotto_object)

# Accessors:
cell_metadata <- pDataDT(visium_giotto_object)
feature_metadata <- fDataDT(visium_giotto_object)

# calculate percentage of mitochondrial content per spot
mitochondrial_genes <- grep('^mt-', visium_giotto_object@feat_ID$rna, value = T)
visium_giotto_object <- addFeatsPerc(gobject = visium_giotto_object,
                             feats = mitochondrial_genes, 
                             vector_name = 'mito')

```

View the gene and cells statistics
```{r}
head(fDataDT(visium_giotto_object)) 
head(pDataDT(visium_giotto_object))
```
Visualise the number of features/genes and mitochondrial genes per cell

```{r}

# Visualize number of genes and mitochondrial content per spot
spatPlot2D(gobject = visium_giotto_object,
           show_image = TRUE,
           point_alpha = 1,
           cell_color = 'nr_feats', color_as_factor = F,
           coord_fix_ratio = 1)

spatPlot2D(gobject = visium_giotto_object,
           show_image = TRUE,
           point_alpha = 1,
           cell_color = 'mito', color_as_factor = F,
           coord_fix_ratio = 1)
```

## Adjust for technical co-founders
Adjust expression matrix for known batch effects or technological covariates. In this case use the mitochondrial genes as a covariate and regress the number of features detected per cell, this means that these covariates will not affect further analyses.
```{r}
visium_giotto_object <- adjustGiottoMatrix(gobject = visium_giotto_object, 
                                   covariate_columns = c("nr_feats", "mito"), 
                                   update_slot = "custom")
```



# Dimension reduction
Prior to running cluster analysis it is best to do feature selection to retain the most informative features/genes to optimise signal to noise ratios by calculating the highly variable featues (HVF). There are three implemented 
```{r}
visium_giotto_object <- calculateHVF(gobject = visium_giotto_object, 
                             method = "cov_groups",
                             save_plot = TRUE)
# visium_giotto_object <- calculateHVF(gobject = visium_giotto_object,
#                              method = "cov_loess",
#                              save_plot = TRUE)
# visium_giotto_object <- calculateHVF(gobject = visium_giotto_object,
#                              method = "var_p_resid",
#                              save_plot = TRUE)

# PCA on expression values 
gene_metadata = fDataDT(visium_giotto_object)
featgenes = gene_metadata[hvf == 'yes' & perc_cells > 3 & mean_expr_det > 0.4]$feat_ID

## run PCA on expression values (default)
visium_giotto_object <- runPCA(gobject = visium_giotto_object,
                       feats_to_use = featgenes)
plotPCA(gobject = visium_giotto_object)
screePlot(gobject = visium_giotto_object, ncp = 30)
```

```{r}
dimPlot2D(gobject = visium_giotto_object,dim_reduction_to_use = "pca")
```

tSNE
```{r}
visium_giotto_object <- runtSNE(visium_giotto_object, 
                                dimensions_to_use = 1:7)
plotTSNE(gobject = visium_giotto_object)
```

UMAP
```{r}
visium_giotto_object <- runUMAP(visium_giotto_object, 
                        dimensions_to_use = 1:7)
plotUMAP(gobject = visium_giotto_object)
```


# Clustering
Before clustering create a nearest network based on similar gene expression. sNN is the default algorithm can also use kNN.
```{r}
visium_giotto_object <- createNearestNetwork(gobject = visium_giotto_object, 
                                     type = "sNN",
                                     dimensions_to_use = 1:10, 
                                     k = 30)
```

Cells can be clustered in Giotto using k-means, Leiden, or Louvain clustering.

## K-means clustering
```{r}
visium_giotto_object <- doKmeans(gobject = visium_giotto_object, 
                         dim_reduction_to_use = 'pca')
```

## Leiden clustering
Increase the resolution to increase the number of clusters
```{r}
visium_giotto_object <- doLeidenCluster(gobject = visium_giotto_object, 
                                resolution = 0.4, 
                                n_iterations = 1000,
                                name = "leiden_0.4_1000")

plotUMAP(gobject = visium_giotto_object,
         cell_color = 'leiden_0.4_1000', 
         show_NN_network = T, 
         point_size = 2.5)
```

## Louvain clustering
Increase the resolution to increase the number of clusters

!! DO NOT USE: Throws an "Not a graph object" error. 
```{r}
# visium_giotto_object <- doLouvainCluster(gobject = visium_giotto_object, 
#                                 resolution = 0.4, 
#                                 version = "community",
#                                 name = "louvain_clus")
# 
# plotUMAP(gobject = visium_giotto_object,
#          cell_color = 'leiden_0.4_1000', 
#          show_NN_network = T, 
#          point_size = 2.5)
```

Clusters of interest can be further sub-clustered. Choose the clusters to be sub-clustered with the selected_clusters argument.

WARNING: Takes a long time!!!ÃŸ
```{r}
## Leiden subclustering for specified clusters
# visium_giotto_object <- doLeidenSubCluster(gobject = visium_giotto_object,
#                              cluster_column = 'leiden_0.4_1000',
#                              resolution = 0.2,
#                              k_neighbors = 10,
#                              hvf_param = list(method = 'cov_loess',
#                                               difference_in_cov = 0.1),
#                              pca_param = list(expression_values = 'normalized',
#                                               scale_unit = F,
#                                               center = F),
#                              nn_param = list(dimensions_to_use = 1:5),
#                              selected_clusters = c(5, 6, 7),
#                              name = 'sub_leiden_clus_select')

#Plot a UMAP to visualize sub-clusters
# plotUMAP(gobject = visium_giotto_object, cell_color = 'sub_leiden_clus_select', show_NN_network = T)
```



# Differentially expressed genes
Now we have clustered the cells, we need to find differentially expressed genes between the identified named clusters, in this case Leiden clusters.
**method** - Method to use fror the the detection of differentially expressed features, can select "scran", "gini" or "mast".
**expression_values** - Feature expession values to use, select either "normalized", "scaled" or "custom".

```{r}
ST_scran_markers_subclusters <- findMarkers_one_vs_all(gobject = visium_giotto_object, 
                                                      method = "scran",
                                                      expression_values ="normalized", 
                                                      cluster_column = "leiden_0.4_1000")
```

Correlation between top selected marker genes and identified clusters
```{r}
ST_top5genes <- ST_scran_markers_subclusters[, head(.SD, 3),by = "cluster"]$feats 

plotMetaDataHeatmap(gobject = visium_giotto_object,
                    selected_feats = ST_top5genes,
                    metadata_cols = c("leiden_0.4_1000"))
```


# Spatial dimension plot
```{r}
# spatial and dimension plots
spatial_dim_plot <-spatDimPlot(gobject = visium_giotto_object, 
                               cell_color = 'leiden_0.4_1000',
                               dim_point_size = 2, 
                               spat_point_size = 2.5)

spatial_dim_plot 
```

```{r}
spatDimPlot(gobject = visium_giotto_object, 
            cell_color = 'nr_feats', 
            color_as_factor = F,
            dim_point_size = 2, 
            spat_point_size = 2.5)
```

```{r}
# dimension plots grouped by cluster
spatPlot2D(visium_giotto_object, 
           cell_color = 'leiden_0.4_1000',
           coord_fix_ratio = 1)
```

```{r}
spatPlot2D(visium_giotto_object, 
           cell_color = 'leiden_0.4_1000',
           group_by = 'leiden_0.4_1000', 
           coord_fix_ratio = 1,
           cow_n_col = 6, 
           show_legend = F,
           save_param = list(base_width = 14, 
                             base_height = 14))
```

```{r}
spatPlot2D(visium_giotto_object, cell_color = 'leiden_0.4_1000',
           select_cell_groups = c('8','4'), coord_fix_ratio = 1, show_other_cells = TRUE,
           cell_color_code = c( "8" = "red","4" = "green4"), other_cell_color = "grey", other_point_size = 1.5,
           save_param = list(base_width = 7, base_height = 7))
```

```{r}
# create and show subset, use the coordiantes from plot above
DG_subset = subsetGiottoLocs(visium_giotto_object,
                             x_max = 1000, x_min = 500,
                             y_max = -800, y_min = -1200,
                             return_gobject = TRUE)

spatDimPlot(gobject = DG_subset, cell_color = "leiden_0.4_1000", spat_point_size = 5)
```

# Marker Gene Detection
Gini markers - Jiang, L., Chen, H., Pinello, L. et al. GiniClust: detecting rare cell types from single-cell gene expression data with Gini index. Genome Biol 17, 144 (2016). https://doi.org/10.1186/s13059-016-1010-4

```{r}
gini_markers_subclusters = findMarkers_one_vs_all(gobject = visium_giotto_object,
                                                  method = 'gini',
                                                  expression_values = 'normalized',
                                                  cluster_column = 'leiden_0.4_1000',
                                                  min_feats = 20,
                                                  min_expr_gini_score = 0.5,
                                                  min_det_gini_score = 0.5)
topgenes_gini = gini_markers_subclusters[, head(.SD, 2), by = 'cluster']$feats

# violinplot
violinPlot(visium_giotto_object, feats = unique(topgenes_gini), cluster_column = 'leiden_0.4_1000',
           strip_text = 8, strip_position = 'right',
           save_param = list(base_width = 5, base_height = 20))
```

```{r}
# cluster heatmap
plotMetaDataHeatmap(visium_giotto_object, selected_feats = unique(topgenes_gini),
                    metadata_cols = c('leiden_0.4_1000'),
                    x_text_size = 10, y_text_size = 10)
```

```{r}
# umap plots
dimFeatPlot2D(visium_giotto_object, expression_values = 'scaled',
              feats = gini_markers_subclusters[, head(.SD, 1), by = 'cluster']$feats,
              cow_n_col = 4, point_size = 0.75,
              save_param = list(base_width = 8, base_height = 8))
```


Scran marker detection
```{r}
# Scran Markers
scran_markers_subclusters = findMarkers_one_vs_all(gobject = visium_giotto_object,
                                                   method = 'scran',
                                                   expression_values = 'normalized',
                                                   cluster_column = 'leiden_0.4_1000')
topgenes_scran = scran_markers_subclusters[, head(.SD, 2), by = 'cluster']$feats

# violinplot
violinPlot(visium_giotto_object, feats = unique(topgenes_scran), cluster_column = 'leiden_0.4_1000',
           strip_text = 10, strip_position = 'right',
           save_param = list(base_width = 5, base_height = 20))
```

```{r}
# cluster heatmap
plotMetaDataHeatmap(visium_giotto_object, selected_feats = topgenes_scran,
                    metadata_cols = c('leiden_0.4_1000'))
```

```{r}
# umap plots
dimFeatPlot2D(visium_giotto_object, expression_values = 'scaled',
              feats = scran_markers_subclusters[, head(.SD, 1), by = 'cluster']$feats,
              cow_n_col = 3, point_size = 1,
              save_param = list(base_width = 8, base_height = 8))

pDataDT(visium_giotto_object)
```

Cell type annotation
PAGE enrichment

Create lists of cell signature genes
```{r}
# Granule neurons
gran_markers <- c("Nr3c2", "Gabra5", "Tubgcp2", "Ahcyl2",
                 "Islr2", "Rasl10a", "Tmem114", "Bhlhe22", 
                 "Ntf3", "C1ql2")

# Oligo dendrocytes
oligo_markers <- c("Efhd1", "H2-Ab1", "Enpp6", "Ninj2",
                  "Bmp4", "Tnr", "Hapln2", "Neu4",
                  "Wfdc18", "Ccp110")        

# Di mesenchephalon
di_mesench_markers <- c("Cartpt", "Scn1a", "Lypd6b",  "Drd5",
                       "Gpr88", "Plcxd2", "Cpne7", "Pou4f1",
                       "Ctxn2", "Wnt4")

# Create a binary matrix of cell signatures
signature_matrix <- makeSignMatrixPAGE(sign_names = c('Granule_neurons',
                                                     'Oligo_dendrocytes',
                                                     'di_mesenchephalon'),
                                      sign_list = list(gran_markers,
                                                       oligo_markers,
                                                       di_mesench_markers))

sign_matrix_path <- system.file("extdata", "sig_matrix.txt", package = 'Giotto')

brain_sc_markers <- data.table::fread(sign_matrix_path)

sig_matrix <- as.matrix(brain_sc_markers[,-1]); rownames(sig_matrix) = brain_sc_markers$Event
```


```{r}
# Enrichment test with PAGE

# runSpatialEnrich() can also be used as a wrapper for all currently provided enrichment options
visium_giotto_object = runPAGEEnrich(gobject = visium_giotto_object, sign_matrix = sig_matrix)

# heatmap of enrichment versus annotation (e.g. clustering result)
cell_types = colnames(sig_matrix)
plotMetaDataCellsHeatmap(gobject = visium_giotto_object,
                         metadata_cols = 'leiden_0.4_1000',
                         value_cols = cell_types,
                         spat_enr_names = 'PAGE',
                         x_text_size = 8, 
                         y_text_size = 8)
```

```{r}
# Visualizations
cell_types_subset = colnames(sig_matrix)[1:10]
spatCellPlot(gobject = visium_giotto_object, 
             spat_enr_names = 'PAGE',
             cell_annotation_values = cell_types_subset,
             cow_n_col = 4,coord_fix_ratio = NULL, point_size = 0.75)
```

```{r}
cell_types_subset = colnames(sig_matrix)[11:20]
spatCellPlot(gobject = visium_giotto_object, spat_enr_names = 'PAGE', 
             cell_annotation_values = cell_types_subset, cow_n_col = 4,
             coord_fix_ratio = NULL, point_size = 0.75, 
             save_param = list(save_name="7_c_spatcellplot_2"))
```


```{r}
# download data to results directory ####
# if wget is installed, set method = 'wget'
# if you run into authentication issues with wget, then add " extra = '--no-check-certificate' "

getSpatialDataset(dataset = 'scRNA_mouse_brain', directory = results_folder)

sc_expression <- paste0(results_folder, "/brain_sc_expression_matrix.txt.gz")
sc_metadata <- paste0(results_folder,"/brain_sc_metadata.csv")

giotto_SC <- createGiottoObject(
  expression = sc_expression,
  instructions = instrs
)

giotto_SC <- addCellMetadata(giotto_SC,
                             new_metadata = data.table::fread(sc_metadata))

giotto_SC <- normalizeGiotto(giotto_SC)
```

```{r}
# 1.2 [shortcut] fully pre-prepared matrix for all cell types
sign_matrix_path = system.file("extdata", "sig_matrix.txt", package = 'Giotto')
brain_sc_markers = data.table::fread(sign_matrix_path)
PAGE_matrix_2 = as.matrix(brain_sc_markers[,-1])
rownames(PAGE_matrix_2) = brain_sc_markers$Event


# ---

# Make PAGE matrix from single cell dataset
markers_scran = findMarkers_one_vs_all(gobject=giotto_SC, method="scran",
                                       expression_values="normalized", cluster_column = "Class", min_feats=3)
top_markers <- markers_scran[, head(.SD, 10), by="cluster"]
celltypes<-levels(factor(markers_scran$cluster))
sign_list<-list()
for (i in 1:length(celltypes)){
  sign_list[[i]]<-top_markers[which(top_markers$cluster == celltypes[i]),]$feats
}

PAGE_matrix_3 = makeSignMatrixPAGE(sign_names = celltypes,
                                   sign_list = sign_list)

# Enrichment test with PAGE

# runSpatialEnrich() can also be used as a wrapper for all currently provided enrichment options
visium_giotto_object = runPAGEEnrich(gobject = visium_giotto_object, sign_matrix = PAGE_matrix_2)

# 1.5 heatmap of enrichment versus annotation (e.g. clustering result)
cell_types_PAGE = colnames(PAGE_matrix_2)
plotMetaDataCellsHeatmap(gobject = visium_giotto_object,
                         metadata_cols = 'leiden_clus',
                         value_cols = cell_types_PAGE,
                         spat_enr_names = 'PAGE',
                         x_text_size = 8,
                         y_text_size = 8)
```

```{r}
# Visualizations
spatCellPlot2D(gobject = visium_giotto_object,
               spat_enr_names = 'PAGE',
               cell_annotation_values = cell_types_PAGE[1:2],
               cow_n_col = 2,coord_fix_ratio = 1, point_size = 1.25, show_legend = T)
```

```{r}
spatDimCellPlot2D(gobject = visium_giotto_object,
                  spat_enr_names = 'PAGE',
                  cell_annotation_values = cell_types_PAGE[1:2],
                  cow_n_col = 1, spat_point_size = 1,
                  plot_alignment = 'horizontal',
                  save_param = list(base_width=7, base_height=10))
```

```{r}
RData.name <- "giotto_spatial_transcriptomics"
save.image(RData.name)
```


```{r}
sessionInfo()
```

